finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/BufferedStream.scala

```diff
@@ -0,0 +1,197 @@
+package com.twitter.finagle.buoyant.h2
+
+import com.twitter.conversions.storage._
+import com.twitter.concurrent.AsyncQueue
+import com.twitter.finagle.buoyant.h2.BufferedStream.{RefCountedDataFrame, RefCountedFrame, RefCountedTrailersFrame, State}
+import com.twitter.finagle.util.AsyncLatch
+import com.twitter.io.Buf
+import com.twitter.util._
+import java.util.concurrent.atomic.AtomicBoolean
+import scala.collection.mutable
+import scala.util.control.NoStackTrace
+
+/**
+ * A BufferedStream reads Frames from an underlying Stream and stores them in a buffer.
+ * BufferedStream itself is not a Stream but child Streams can be created from it by calling
+ * fork().  Child Streams will first be offered all of the buffered Frames and then all Frames
+ * read from the underlying Stream will also offered to the child Streams.
+ *
+ * When the buffer becomes full or when discardBuffer() is called, the buffer will be discarded
+ * and no new child Streams may be created.  However, child Streams that had previously been
+ * created will continue to receive Frames as they are read from the underlying Stream.
+ *
+ * The Frames that are stored in the buffer and offered to the child Streams are refcounted and
+ * the underlying Frame will only be released once each child Stream releases the Frame and the
+ * buffer itself releases the Frame.  Threrefore you should always call discardBuffer on a
+ * BufferedStream before it leaves scope.
+ */
+class BufferedStream(underlying: Stream, bufferCapacity: Long = 8.kilobytes.bytes) {
```

Comment:
technically, according to the h2 spec, the window size may change.  it may be worth modeling capacity as a Var.  but i wouldn't make this a blocker.  Also, i'd default this to 65535 bytes to match the rfc default window size